scio-snowflake/src/main/scala/com/spotify/scio/snowflake/SnowflakeIO.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.snowflake
+
+import com.spotify.scio.ScioContext
+import com.spotify.scio.coders.{Coder, CoderMaterializer}
+import com.spotify.scio.io.{EmptyTap, EmptyTapOf, ScioIO, Tap, TapT}
+import com.spotify.scio.values.SCollection
+import kantan.csv.{RowDecoder, RowEncoder}
+import org.apache.beam.sdk.io.snowflake.SnowflakeIO.{CsvMapper, UserDataMapper}
+import org.apache.beam.sdk.io.{snowflake => beam}
+
+object SnowflakeIO {
+
+  private[snowflake] def dataSourceConfiguration(connectionOptions: SnowflakeConnectionOptions) = {
+
+    val datasourceInitial = beam.SnowflakeIO.DataSourceConfiguration
+      .create()
+
+    val datasourceWithAuthent = connectionOptions.authenticationOptions match {
+      case SnowflakeUsernamePasswordAuthenticationOptions(username, password) =>
+        datasourceInitial.withUsernamePasswordAuth(username, password)
+      case SnowflakeKeyPairAuthenticationOptions(username, privateKeyPath, None) =>
+        datasourceInitial.withKeyPairPathAuth(username, privateKeyPath)
+      case SnowflakeKeyPairAuthenticationOptions(username, privateKeyPath, Some(passphrase)) =>
+        datasourceInitial.withKeyPairPathAuth(username, privateKeyPath, passphrase)
+      case SnowflakeOAuthTokenAuthenticationOptions(token) =>
+        datasourceInitial.withOAuth(token)
+    }
+
+    val datasourceBeforeSchema = datasourceWithAuthent
+      .withServerName(connectionOptions.serverName)
+      .withDatabase(connectionOptions.database)
+      .withRole(connectionOptions.role)
+      .withWarehouse(connectionOptions.warehouse)
+
+    connectionOptions.schema
+      .map(schema => datasourceBeforeSchema.withSchema(schema))
+      .getOrElse(datasourceBeforeSchema)
+  }
+
+  private[snowflake] def buildCsvMapper[T](rowDecoder: RowDecoder[T]): CsvMapper[T] =
+    new CsvMapper[T] {
+      override def mapRow(parts: Array[String]): T = {
+        val unsnowedParts = parts.map {
+          case "\\N" => "" // needs to be mapped to an Option
```

Comment:
I never used Snowflake. Can you give some context for this case ?

Reply:
`SnowflakeIO` from Apache Beam uses `COPY` from Snowflake, that simply exports to CSV, that is then read from storage. And this copy uses `\N` the represent `null`.

Thing is, kantan maps empty string with `None`, so we provide it.